File: packages/core/src/core/divergence.py
"""El detector de divergencia. La pieza que puntúa en Adaptación.

Cada tick se reevalúan las suposiciones del plan vigente contra el estado real:
`divergence = suposiciones_rotas_ponderadas / total_ponderado`.

Por encima de 0,25, replan. Además hay dos disparadores inmediatos: violación de
restricción dura, y llegada de un hecho con `severity: critical`.

El valor va al dashboard como una línea que sube y cruza el umbral justo antes del
banner rojo.
"""

from contracts.calls import Fact
from contracts.plan import DIVERGENCE_THRESHOLD, PlanContext
from contracts.world import WorldState
from core.belief import road_of

WIND_TOLERANCE_DEG = 20.0
"""Cuánto puede virar el rumbo del viento antes de contar la suposición como rota.
El coste de ruta apenas cambia con un viraje pequeño; solo un giro real invalida el
plan. Sin tolerancia, cualquier ruido de un grado dispara replans en cadena."""


def divergence(state: WorldState, ctx: PlanContext) -> tuple[float, list[str]]:
    """(valor, claves de las suposiciones rotas). Puro.

    `divergence = suposiciones_rotas_ponderadas / total_ponderado`. Las suposiciones
    no evaluables (clave que no se sabe leer, arista desaparecida) se saltan: ni suman
    al numerador ni al denominador, para no inflar falsos positivos."""
    broken_weight = 0.0
    total_weight = 0.0
    broken: list[str] = []

    for a in ctx.assumptions:
        actual = evaluate_assumption(state, a.key)
        if actual is None:
            continue
        total_weight += a.weight
        if _is_broken(a.key, actual, a.expected):
            broken_weight += a.weight
            broken.append(a.key)

    if total_weight == 0.0:
        return 0.0, []
    return broken_weight / total_weight, broken


def should_replan(
    value: float, violations_hard: int, critical_facts: list[Fact]
) -> tuple[bool, str]:
    """(bandera, motivo). El motivo va al banner y al prompt del planner tal cual.

    Si esto devuelve False no se llama al modelo, y eso es lo que mantiene el coste
    y la latencia bajo control."""
    if violations_hard > 0:
        return True, "restricción dura violada"
    if critical_facts:
        keys = ", ".join(f.key for f in critical_facts)
        return True, f"hecho crítico: {keys}"
    if value > DIVERGENCE_THRESHOLD:
        return True, f"divergencia {value:.2f} > {DIVERGENCE_THRESHOLD}"
    return False, ""


def evaluate_assumption(state: WorldState, key: str) -> str | float | bool | None:
    """El valor actual de una suposición ("road:wp_sur_03-wp_sur_04:open").

    Espejo de las claves que emite `solver.build_context`. `None` = clave que no se
    sabe leer o cuyo objeto ya no existe (suposición no evaluable)."""
    seg = key.split(":")

    if seg[0] == "road" and len(seg) == 3 and seg[2] == "open":
        edge = road_of(state.roads, seg[1])
        return None if edge is None else not edge.cut

    if key == "wind:bearing_deg":
        return state.wind.bearing_deg
    if key == "wind:speed":
        return state.wind.speed

    if seg[0] == "cell" and len(seg) == 3 and seg[2] == "state":
        cell = state.cells.get(seg[1])
        return None if cell is None else cell.state

    # Fallback: el último hecho con esa clave, si alguno entró al estado.
    for fact in reversed(state.facts):
        if fact.key == key:
            return fact.value
    return None


def _is_broken(
    key: str, actual: str | float | bool, expected: str | float | bool
) -> bool:
    """`actual != expected`. Los rumbos (grados) llevan tolerancia circular; el resto
    de floats, tolerancia relativa; bools y strings, igualdad exacta."""
    if isinstance(expected, bool) or isinstance(actual, bool):
        return actual != expected
    if isinstance(expected, (int, float)) and isinstance(actual, (int, float)):
        if key.endswith("bearing_deg"):
            return _angular_gap(float(actual), float(expected)) > WIND_TOLERANCE_DEG
        return abs(float(actual) - float(expected)) > 1e-6 * max(
            1.0, abs(float(expected))
        )
    return actual != expected


def _angular_gap(a: float, b: float) -> float:
    """Diferencia mínima entre dos ángulos en grados, con el wrap 0–360."""
    d = abs(a - b) % 360.0
    return min(d, 360.0 - d)

File: apps/dashboard/src/sim2d/engine/divergence.ts
// Cuánto se ha alejado el mundo de lo que el plan daba por cierto.
//
// Puerto de `packages/core/src/core/divergence.py`. Es el número que decide cuándo se
// tira el plan, y el que cruza 0,25 en el gráfico del dashboard.
//
// Una suposición que no se puede evaluar NO cuenta: ni en el numerador ni en el
// denominador. Si contara como rota, un plan viejo con claves que ya no existen daría
// divergencia 1 y replanificaría para siempre.
import { DIVERGENCE_THRESHOLD } from '../../types'
import type { Assumption, PlanContext } from '../../types'

import { angularGap } from './geom'
import { latestFact, type SimState } from './state'

export const WIND_TOLERANCE_DEG = 20.0

export interface Term {
  key: string
  expected: string | number | boolean
  actual: string | number | boolean | null
  weight: number
  /** `null` = no evaluable: no suma ni en roto ni en total. */
  broken: boolean | null
}

export interface DivergenceResult {
  value: number
  terms: Term[]
  broken: string[]
  brokenWeight: number
  totalWeight: number
}

function evaluate(state: SimState, a: Assumption): { actual: Term['actual']; broken: boolean | null } {
  const road = /^road:(.+):open$/.exec(a.key)
  if (road?.[1] !== undefined) {
    const edge = state.roads.get(`road:${road[1]}`) ?? state.roads.get(road[1])
    if (!edge) return { actual: null, broken: null }
    const open = !edge.cut
    return { actual: open, broken: open !== a.expected }
  }
  if (a.key === 'wind:bearing_deg') {
    const actual = state.wind.bearing_deg
    const expected = Number(a.expected)
    // Circular, con tolerancia: 350° y 10° distan 20°, no 340°.
    return { actual, broken: angularGap(actual, expected) > WIND_TOLERANCE_DEG }
  }
  if (a.key === 'wind:speed') {
    const actual = state.wind.speed
    return { actual, broken: !closeEnough(actual, Number(a.expected)) }
  }
  const cell = /^cell:(.+):state$/.exec(a.key)
  if (cell?.[1] !== undefined) {
    const c = state.cells.get(cell[1])
    const actual = c?.state ?? 'intact'
    return { actual, broken: actual !== a.expected }
  }
  const fact = latestFact(state, a.key)
  if (!fact) return { actual: null, broken: null }
  if (typeof fact.value === 'number' && typeof a.expected === 'number') {
    return { actual: fact.value, broken: !closeEnough(fact.value, a.expected) }
  }
  return { actual: fact.value, broken: fact.value !== a.expected }
}

function closeEnough(a: number, b: number): boolean {
  return Math.abs(a - b) <= 1e-6 * Math.max(1, Math.abs(b))
}

export function divergence(state: SimState, context: PlanContext): DivergenceResult {
  const terms: Term[] = []
  let brokenWeight = 0
  let totalWeight = 0
  const broken: string[] = []

  for (const a of context.assumptions) {
    const { actual, broken: isBroken } = evaluate(state, a)
    terms.push({ key: a.key, expected: a.expected, actual, weight: a.weight, broken: isBroken })
    if (isBroken === null) continue
    totalWeight += a.weight
    if (isBroken) {
      brokenWeight += a.weight
      broken.push(a.key)
    }
  }

  return {
    value: totalWeight > 0 ? brokenWeight / totalWeight : 0,
    terms,
    broken,
    brokenWeight,
    totalWeight,
  }
}

export interface ReplanDecision {
  flag: boolean
  reason: string
}

/** El mismo orden que `divergence.should_replan`: una violación dura gana a un hecho
 *  crítico, y los dos ganan al umbral. El orden importa porque es el que se enseña. */
export function shouldReplan(
  value: number,
  hardViolations: number,
  criticalFacts: readonly string[],
): ReplanDecision {
  if (hardViolations > 0) return { flag: true, reason: 'restricción dura violada' }
  if (criticalFacts.length > 0) return { flag: true, reason: `hecho crítico: ${criticalFacts.join(', ')}` }
  if (value > DIVERGENCE_THRESHOLD) {
    return {
      flag: true,
      reason: `divergencia ${value.toFixed(2).replace('.', ',')} > ${String(DIVERGENCE_THRESHOLD).replace('.', ',')}`,
    }
  }
  return { flag: false, reason: '' }
}
